corehq/motech/repeaters/tasks.py

```diff
@@ -72,65 +64,103 @@ def delete_old_request_logs():
-                if not _soft_assert(
-                    datetime.utcnow() < twentythree_hours_later,
-                    "I've been iterating repeat records for 23 hours. I quit!"
-                ):
-                    break
-
+            for repeater in iter_ready_repeaters():
                 metrics_counter("commcare.repeaters.check.attempt_forward")
-                record.attempt_forward_now(is_retry=True)
-            else:
-                iterating_time = datetime.utcnow() - start
-                _soft_assert(
-                    iterating_time < timedelta(hours=6),
-                    f"It took {iterating_time} to iterate repeat records."
-                )
+                process_repeater.delay(repeater.domain, repeater.repeater_id)
+
+                if datetime.utcnow() > twentythree_hours_later:
+                    # Break after process_repeater() to avoid a stale lock
+                    break
     finally:
         check_repeater_lock.release()
 
 
+def iter_ready_repeaters():
+    """
+    Cycles through repeaters (repeatedly ;) ) until there are no more
+    repeat records ready to be sent.
+    """
+    while True:
+        yielded = False
+        for repeater in Repeater.objects.all_ready():
+            if not repeater.domain_can_forward:
+                continue
+
+            if rate_limit_repeater(repeater.domain):
+                repeater.rate_limit()
+                continue
+
+            yielded = True
+            yield repeater
+
+        if not yielded:
+            # No repeaters are ready, or they are rate limited, or their
+            # domains can't forward or are paused.
+            return
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def process_repeater(domain, repeater_id):
+
+    def is_retry(repeat_record):
+        return repeat_record.state == State.Fail
+
+    repeater = Repeater.objects.get(domain=domain, id=repeater_id)
+    repeat_records = repeater.repeat_records_ready[:repeater.num_workers]
+    header_tasks = [
+        retry_process_repeat_record.s(rr.id, rr.domain)
+        if is_retry(rr)
+        else process_repeat_record.s(rr.id, rr.domain)
+        for rr in repeat_records
+    ]
+    chord(header_tasks)(update_repeater.s(repeater.repeater_id))
+
+
+@task(queue=settings.CELERY_REPEAT_RECORD_QUEUE)
+def update_repeater(repeat_record_states, repeater_id):
+    """
+    Determines whether the repeater should back off, based on the
+    results of `fire_repeat_record()` tasks.
+    """
+    repeater = Repeater.objects.get(id=repeater_id)
+    if any(s == State.Success for s in repeat_record_states):
+        # At least one repeat record was sent successfully.
+        repeater.reset_backoff()
+    elif all(s in (State.Empty, None) for s in repeat_record_states):
+        # Nothing was sent. Don't update the repeater.
+        pass
+    else:
+        # All sent payloads failed.
+        repeater.set_backoff()
```

Comment:
Looks like this else block could be executed if all states are either `Pending`, `Failed`, or `Canceled`. I think it makes sense to backoff for `Pending` or `Failed` (which both mean _pending_), but `Cancelled` seems more like the `Success` state and should not cause backoff. Do you agree?

Reply:
Good question, and great point, because `Cancelled` can mean two different things:
* The payload is bad, not the remote endpoint, and has been cancelled. (Repeater should not back off.)
* Sending failed one time too many (`exceeded_max_retries` is `True`), and the payload has been cancelled. (Repeater should back off.)

(The state should never be `Pending`, because to reach this task we much have attempted to send the payload.)